torchrl/collectors/collectors.py

```diff
@@ -795,11 +795,14 @@ def _step_and_maybe_reset(self) -> None:
                 dtype=torch.bool,
             )
             if td_reset.batch_dims:
+                # better cloning here than when passing the td for stacking
+                # cloning is necessary to avoid modifying dones in-place
+                self._tensordict = self._tensordict.clone()
```

Comment:
Can you explain a bit more this? to avoid modifying the done we have to clone all the td?

Reply:
First note that this clone and the other only occur when we reset. For pendulum for instance, that is 1% of the trials (more if you have a lot of envs in the collector obv)
The reason is the same as the one in traj_ids: after this, we index the `_tensordict` with the `_reset` mask and change values in place. If we don't clone, we will modify the td we keep in the cache in rollout, therefore modifying the data at t-1, which should be avoided at all cost

Reply:
For the first note, I think we should not apply that reasoning.
I think batched environments will take over in RL since you can run them parallely on the GPU.
In those envs we will basically be resetting at every step bacause at least 1 out of 60_000 envs will be done.

So it is better to assume that reset is run at every step.

That said, yeah if we can't avoid it there is nothing to do.
But maybe we can only clone some keys of interest if the possible modification will not impact all keys

Reply:
We must copy the data since reset resets the observation at the root for the selected tds. The alternative is to clone at each iteration in rollout, which is far worse
Looking at this we may have a solution:

![image](https://github.com/pytorch/rl/assets/25529882/a146bb35-6941-48c2-9590-1fdded85de46)

I can try it if you want

Reply:
Leaving this as a TODO as it breaks with lazy stacked tds